cdpinger/man/ltspfsd.1
.TH "LTSPFSD" "1" 
.SH "NAME" 
ltspfsd \(em daemon to serve fuse based filesystems over the network 
.SH "SYNOPSIS" 
.PP 
\fBltspfsd\fR [\fB-a \fP]  [\fB-r \fP]  [\fB-d \fP]  
.SH "DESCRIPTION" 
.PP 
This manual page documents briefly the 
\fBltspfsd\fR command. 
 
.PP 
\fBltspfsd\fR is a program that provides access to local devices over the network from a ltspfs client 
.SH "OPTIONS" 
.IP "\fB-a\fP         " 10 
for no authentication. The server will simply accept the XAUTH packet sent down, and always return OK. 
.IP "\fB-r\fP         " 10 
for read only. Specifying this causes the ltspfsd to behave as a readonly server, and won't allow any modification to the media. 
.IP "\fB-d\fP         " 10 
for debug. Server won't fork into the background, and will print messages to stderr. 
.SH "SEE ALSO" 
.PP 
ltspfs (1). 
.SH "AUTHOR" 
.PP 
This manual page was written by Oliver Grawert ogra@ubuntu.com for 
the \fBDebian\fP system (but may be used by others).  Permission is 
granted to copy, distribute and/or modify this document under 
the terms of the GNU General Public License, Version 2 any  
later version published by the Free Software Foundation. 
 
.PP 
On Debian systems, the complete text of the GNU General Public 
License can be found in /usr/share/common-licenses/GPL. 
 
.\" created by instant / docbook-to-man, Fri 25 Aug 2006, 21:38 


cdpinger/ChangeLog
  157 Vagrant Cascadian 2012-02-24
      Add symlinks for Debian/Ubuntu for ltspfsd.rules creation in
      init-ltsp.d.

  156 Vagrant Cascadian 2012-02-24
      Use init-ltsp.d hook for installing udev rules at boot.

  155 Vagrant Cascadian 2012-02-24
      Rename udev rules to ltspfsd.rules, and move them into
      /usr/share/ltspfs/udev, which allows for installing rules at boot.

  154 vagrant@freegeek.org	2011-10-28 {ltspfs-1.0}
      version 1.0

  153 vagrant@freegeek.org	2011-10-28 [merge]
      merge patch from Scott Balneaves to fix 64 bit issues.

  152 Stéphane Graber	2011-08-11 {ltspfs-0.9}
      version 0.9

  151 Stéphane Graber	2011-07-22
      Launchpad: no change commit to trigger a test rebuild

  150 Stéphane Graber	2011-07-07
      ID_FS_LABEL is now usually 'safe'. Patch from gcc (LP: #690969)

  149 Warren Togami	2011-06-16 {ltspfs-0.8}
      version 0.8

  148 Vagrant Cascadian 2011-02-05
      cdpinger: add support for UDF formatted CDs and DVDs.

  147 Stéphane Graber	2010-08-08 {ltspfs-0.7}
      Release 0.7

  146 Vagrant Cascadian 2010-06-25
      init with "ltspfs" rather than simply "test".

  145 Vagrant Cascadian 2010-06-25
      Add example ltspfs mounted hook to issue notifications for
      desktops that do not display icons. Thanks to Petter Reinholdtsen.
      http://bugs.debian.org/575031

  144 Gideon Romm	2010-05-11
      Remove commented-out code.

  143 Gideon Romm	2010-05-11
      Fix cdrom handling.

  142 Stéphane Graber	2010-02-17 {ltspfs-0.6}
      version 0.6

  141 Stéphane Graber	2010-02-17
      Try umounting with -l to avoid some issues

  140 Scott Balneaves	2010-02-17
      Change owner of dir in /media to user, with group of root

  139 Scott Balneaves	2010-02-17
      Top level media directory being created mode 750.  Changed to 700

  138 Gideon Romm	2010-02-10
      Be careful to actually have a pid before executing kill

  137 Gideon Romm	2010-02-10
      Do not sed -i a file that we are reading in from stdin

  136 Vagrant Cascadian 2010-02-10
      allow CDs to have an empty filesystem, as they're likely to have
      a device but no disk inserted. this allows cdpinger to run again.

  135 Vagrant Cascadian 2010-02-10
      cdpinger: report syslog messages for "disk detected" and "tray
      open" messages only when the state changes so they are not reported
      every 3 seconds.

  134 Gideon Romm	2010-02-10
      Fix MODE setting to include add_disc annd remove_disc

  133 Gideon Romm	2010-02-10
      Pass DEVICENAME to add_cdrom_device and remove_cdrom_device
      functions

  132 Gideon Romm	2010-02-10
      Make sure ltspfsd gets launched

  131 Gideon Romm	2010-02-10
      Refactor ltspfs_entry and cdpinger code to better handle CDROMs

  130 Gideon Romm	2010-02-10
      Remove cdpinger pid file when cdrom is removed, since apparently
      cdpinger will commit suicide on its own upon device removal and
      we would like cdpinger to start up again upon reinsertion of the
      cdrom drive.

  129 Gideon Romm	2010-02-09
      Try to avoid multiple parallel calls to pgrep by writing a pid
      file to check first. If the pid file exists, it shell should not
      need to use pgrep in the conditional

  128 Gideon Romm	2010-02-01
      Make delayed-mounter work for LDM localapps

  127 Scott Balneaves	2010-01-30
      Added syslogging.

  126 Gideon Romm	2010-01-29
       Some more optimizations to use shell primitives where possible

  125 Vagrant Cascadian 2010-01-29
      fix call_cdpinger function to use exact matching so that "cdrom"
      doesn't match "cdrom1". thanks gadi!

  124 Gideon Romm	2010-01-27
      Fix up noauth stuff and make local devices work for localapps,
      rdesktop, and kiosk scripts.

  123 Scott Balneaves	2010-01-27
      Changed noauth mode: ltspfs now recognizes LTSPFS_NOAUTH env var,
      and if set, skips auth.  ltspfsd now skips receiving auth packet
      if noauth set.

  122 Stéphane Graber	2010-01-13 {ltspfs-0.5.14}
      release ltspfs 0.5.14

  121 Stéphane Graber	2010-01-13
      Fix cases where the user used for login doesn't match the case of
      the remote user

  120 Scott Balneaves	2010-01-12
      Files are now user-owned only

  119 Marc Gariepy	2009-12-15
      fixing local lowercase username (ldap ltsp-trunk -r 1265)

  118 Vagrant Cascadian 2009-07-17 {ltspfs-0.5.13}
      release ltspfs 0.5.13

  117 Vagrant Cascadian 2009-07-17
      drop delayedmounter from xinitrc.d, as udev scripts should be
      sufficient now.

  116 Vagrant Cascadian 2009-07-06
      relocate delayed-mounter ldm hooks so they run after the user
      is added to /etc/passwd, and before the user is removed from
      /etc/passwd.

  115 Vagrant Cascadian 2009-07-06
      - always run as root, so mounts appear in /media/root - only run
      mount for localapps if user's home dir is mounted via sshfs

  114 Vagrant Cascadian 2009-07-06
      check for ldm-xauth in /var/run, and support newer ldm, with
      Xauthority in a sub-dir.

  113 Vagrant Cascadian 2009-06-21
      fix spelling of "Received" in ltspfsd

  112 Stéphane Graber	2009-06-03 {ltspfs-0.5.12}
      version 0.5.12

  111 Francis Giraldeau 2009-05-27
      remove ltspfs_fstab entry before unmounting device,
      to prevent other processes from mounting it.
      https://bugs.launchpad.net/ubuntu/+source/ltspfsd/+bug/378495

  110 Vagrant Cascadian 2009-04-03 {ltspfs-0.5.11}
      version 0.5.11

  109 Vagrant Cascadian 2009-04-03
      simplify code to handle when mountpoint is already mounted

  108 Gideon Romm	2009-03-11
      Make sure delayed_mounter sets USER=root for local ltspfsmounting

  107 Gideon Romm	2009-03-10
      Do not remove ltspfs mount on "already mounted" errors, as these
      errors may occur silently with multiple ltspfsmounters

  106 Stéphane Graber	2009-03-09 {ltspfs-0.5.10}
      version 0.5.10

  105 Stéphane Graber	2009-03-08

cdpinger/src/ltspfsd.c
/*
 * ltspfsd.c
 * Mainline handling for the Linux Terminal Server Project File Server Daemon.

 * Copyright Scott Balneaves, sbalneav@ltsp.org, 2005, 2006, 2007
 *           Gideon Romm, gadi@ltsp.org, 2009

 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, you can find it on the World Wide
 * Web at http://www.gnu.org/copyleft/gpl.html, or write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
 * MA 02110-1301, USA.

 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <rpc/xdr.h>
#include <signal.h>
#include <errno.h>
#include <string.h>
#include <syslog.h>
#include <fcntl.h>
#include <unistd.h>
#include <pwd.h>
#include <X11/Xlib.h>
#include <X11/Xauth.h>
#include <X11/Xatom.h>
#include "ltspfs.h"
#include "ltspfsd_functions.h"
#include "common.h"

/*
 * Globals.
 */

int debug;                      /* Debug level.  0 = no debug */
int readonly;                   /* If true, make filesystem "read only" */
int noauth;                     /* If true, skip authentication */
int syslogopen;                 /* If true, then log to syslog, else stderr */
char *mountpoint;               /* Point we're "mounted" to */
int authenticated;              /* Mountpoint length */
int mounted;                    /* Automounter status */

/*
 * Signal handler function for SIGCHLD
 */

void sig_chld(int signo)
{
    pid_t pid;
    int stat;
    (void) signo;

    pid = wait(&stat);                           /* reap our child process */
    (void) pid;
}

/*
 * Signal handler function for SIGTERM
 */

void sig_term(int signo)
{
    (void) signo;
    info("Received SIGTERM, shutting down\n");
    exit(OK);
}

void timeout(void)
{
    exit(0);
}

/*
 * mainline
 */

int main(int argc, char *argv[])
{

    int sockfd, newsockfd, childpid;
    struct sockaddr_in cli_addr;
    socklen_t clilen;
    int c;

    authenticated = 0;
    mounted = 0;
    mountpoint = NULL;

    /*
     * We can begin to parse command line options.
     */

    debug = readonly = opterr = noauth = 0;

    while ((c = getopt(argc, argv, "rad")) != -1) {
        switch (c) {
        case 'r':
            readonly++;
            break;
        case 'a':
            noauth++;
            break;
        case 'd':
            debug++;
            break;
        case '?':
            fprintf(stderr, "Unknown option character `\\x%x'.\n", optopt);
            return 1;
        default:
            break;
        }
    }

    /*
     * Some initialization work.  We call daemon(), which will chdir to /
     * so that the daemon won't lock any filesystems. The daemon() call also
     * closes STDIN, STDOUT, and STDERR, and forks.  Then, set up our umask,
     * and redirect the SIGCHLD and SIGTERM signals to handlers.
     */

    if (!debug) {                                /* if !debug, background */
        if (!daemon(0, 0)) {
            signal(SIGCHLD, sig_chld);
            signal(SIGTERM, sig_term);
        }
    }

    umask(0);

    /*
     * open our logging.
     */

    if (!debug) {
        openlog(argv[0], LOG_PID, LOG_USER);
        syslogopen++;
    } else
        setbuf(stderr, NULL);

    /*
     * Open our socket
     */

    sockfd = bindsocket(SERVER_PORT);

    /*
     * Log informational message to indicate program starting
     */

    info("Program started\n");

    /*
     * Calculate the length of the cli_addr structure (needed for the accept)
     */

    clilen = sizeof(cli_addr);

    for (;;) {

        /*
         * Accept incoming connection
         */

        newsockfd = accept(sockfd, (struct sockaddr *) &cli_addr, &clilen);
        if (newsockfd < 0)
            error_die("Can't accept connection");

        /* 
         * Added this part to make local connections ignore authentication
         */
        char destination_buffer[INET_ADDRSTRLEN]; 
        inet_ntop(AF_INET, (void *)&cli_addr.sin_addr, destination_buffer, sizeof(destination_buffer)); 
        if (!strcmp("127.0.0.1", destination_buffer)) { 
            noauth=1; 
        } else {
            noauth=0;
        }

        if (!debug)                              /* In debug mode, don't fork */
            childpid = fork();
        else
            childpid = 0;

        if (childpid < 0)
            error_die("fork error");             /* If fork fails, die */
        else if (childpid == 0) {
            close(sockfd);                       /* Pass work off to child */
            handle_connection(newsockfd);
            exit(OK);                            /* exit child, SIGCHLD sent */
        }
        close(newsockfd);                        /* parent */
    }
    return OK;                                   /* return 0 from main */
}

/*
 * handle_connection is where the "work" gets done.  Anything you want to "do"
 * as a network daemon should be done here.
 *
 * Dispatch ltspfs calls to ltspfs_dispatch.
 */

void handle_connection(int sockfd)
{
    XDR in;
    int n;
    int i, q;
    char line[LTSP_MAXBUF];
    char *lineptr;
    fd_set set;                 /* For select */
    struct timeval automount_timeout;   /* Timeout */
    int nleft, nread;
    int r;


    for (;;) {
        FD_ZERO(&set);
        FD_SET(sockfd, &set);
        xdrmem_create(&in, line, LTSP_MAXBUF, XDR_DECODE);
        lineptr = line;

        /*
         * Sigh.  I really hate to have to re-duplicate most of the readn
         * functionality here, but it's simply cleaner to not butcher the _readn
         * and _writen primitives just to handle automount timeouts. Soooo....
         */

        automount_timeout.tv_sec = AUTOMOUNT_TIMEOUT;
        automount_timeout.tv_usec = 0;
        r = select(FD_SETSIZE, &set, NULL, NULL, &automount_timeout);
        if (r < 0)
            error_die("handle_connection: select error\n");
        else if (r == 0) {
            if (mounted)
                am_umount(mountpoint);           /* this will return */
            continue;                            /* Back to the top of loop */
        }

        nleft = BYTES_PER_XDR_UNIT;

        while (nleft > 0) {
            nread = read(sockfd, lineptr, nleft);
            if (nread < 0)
                error_die("handle_connection: readline error\n");
            else if (nread == 0)
                return;                          /* EOF, connection closed */

            nleft -= nread;
            lineptr += nread;
        }

        xdr_int(&in, &i);
        if (debug)
            info("Packet length: %d\n", i);
        n = readn(sockfd, lineptr, (i - BYTES_PER_XDR_UNIT));
        if (n == 0)
            return;                              /* connection closed */
        else if (n < 0)
            error_die("handle_connection: readline error\n");

        /*
         *  OK We've got our command line, pass this off to ltspfs_dispatch
         */

        if (debug) {
            info("Packet buffer: ");
            for (q = 0; q < i; q++)
                info("%x", line[q]);
            info("\n");
        }

        ltspfs_dispatch(sockfd, &in);

        /*
         * Well, we're back.  Reset our position to the beginning.
         */
        xdr_destroy(&in);
    }
}
